[LV] Prevent query the computeCost() when VF=1 in emitInvalidCostRemarks().

ElvisWang123 opened the pull request:
We should only query the computeCost() when the VF is vector.

ElvisWang123 committed:
[LV] Prevent query the computeCost() when VF=1 in emitInvalidCostRemarks().

ElvisWang123 commented:
Soft ping!

david-arm reviewed (approved):
LGTM! Seems sensible.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4372,6 +4372,9 @@ void LoopVectorizationPlanner::emitInvalidCostRemarks(
   SmallVector<RecipeVFPair> InvalidCosts;
   for (const auto &Plan : VPlans) {
     for (ElementCount VF : Plan->vectorFactors()) {
+      if (VF.isScalar())
```
Could you add a comment here why we skip scalar VFs? Presumably because we don’t expect them to have invalid costs?

ElvisWang123 replied:
Updated comments.

Currently, we cannot query the vplan-based cost model with scalar VF because we use lots of `cast<VectorType>(toVectorTy(SrcTy, VF))` in `computeCost()`.

ElvisWang123 committed:
Add comments to explain why we skip scalar cost.

ElvisWang123 committed:
Add test case.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -4499,6 +4499,12 @@ void LoopVectorizationPlanner::emitInvalidCostRemarks(
   for (const auto &Plan : VPlans) {
     for (ElementCount VF : Plan->vectorFactors()) {
+      // The VPlan-based cost model is designed for computing vector cost.
+      // Quering VPlan-based cost model with scarlar VF will cause some error
```
nit: `// Querying VPlan-based cost model with a scalar VF will cause some errors`

ElvisWang123 replied:
Fixed, thanks!

david-arm commented on llvm/test/Transforms/LoopVectorize/RISCV/remark-reductions.ll:
```diff
@@ -0,0 +1,65 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt < %s -mtriple=riscv64 -mattr=+v -p loop-vectorize -pass-remarks-analysis=loop-vectorize -S | FileCheck %s
```
This doesn't feel quite right, since you're passing `-pass-remarks-analysis=loop-vectorize` but not testing the output. Given this test is about remarks, I think we should check the output and also ensure we don't see the VF=1 case.

ElvisWang123 replied:
Passing `-pass-remarks-analysis=loop-vectorize` just for  triggering  `emitInvalidCostRemarks()` and  make sure the compiler don't crash. I don't have any ideas about testing the remark output.

Something I can imagine is that using  `--debug-only=loop-vecotorize` to print the cost. But it always using the legacy cost model to calculate/print the scalar cost.

Do you have better suggestion to test this case?

fhahn replied:
You could also check the remark output, assuming there is any? Or of not, check that there is no remark output

ElvisWang123 replied:
Update test, thanks!

ElvisWang123 committed:
!fixup typo and update test case.

Becaue the original test case will not vectorized anymore. Update the
test case the will hit the bug.

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/remark-reductions.ll:
```diff
@@ -0,0 +1,65 @@
+loop:
+  %iv = phi i32 [ 0, %entry ], [ %iv.next, %loop ]
+  %red = phi float [ 0.000000e+00, %entry ], [ %red.next, %loop ]
+  %red.next = fadd float 0.000000e+00, %red
```
this is a no-op reduction, can this also reproduce with a 'real' reductions?

ElvisWang123 replied:
Yes, update test case to real reduction.
https://godbolt.org/z/dj3hrK8fT

ElvisWang123 committed:
!Fixup, using real reduction and checks remark output.

fhahn reviewed (approved):
LGTM, thanks

ElvisWang123 committed:
Merge branch 'main' into fix-missing-VF-check-in-emitRemark-LV
